### ISSUE TITLE
Minibus contrib: escalating money events

### ISSUE THREAD

**Issue by JWJoubert** [matsim-org/matsim-libs#1581]
Hi @neuma, @vsp-gleich 

I am trying to escalate a money event (like a fine, or toll) from the minibus vehicle, all the way back to the association (the agent deciding on how to change the routes and/or vehicles). The [code example](https://github.com/matsim-org/matsim-libs/blob/01ef87abd7e3faca035ead42c3ba27c75d98adca/contribs/minibus/src/main/java/org/matsim/contrib/minibus/RunConfigurableMinibusExample.java) is no longer working.

A `PersonMoneyEvent` seems to only hit the commuter's pocket; not the person driving the minibus ;-)

I then tried doing this using a `LinkEnterEvent` listener:
- from the event I can get the vehicle `Id`;
- I then have to iterative search through the `POperators` and match the vehicle's `Id` with the `POperator`'s `Id`;
- I then reduce the `POperator`'s budget with whatever the fine amount is.

Currently, I do not seem to get the behavioural response I'm expecting.

I do not know if there is an intuitively more accurate way of doing it? Especially using existing `minibus` infrastructure. I was hoping to use something like the [`StageContainer2AgentMoneyEvent`](https://github.com/matsim-org/matsim-libs/blob/01ef87abd7e3faca035ead42c3ba27c75d98adca/contribs/minibus/src/main/java/org/matsim/contrib/minibus/scoring/StageContainer2AgentMoneyEvent.java) or the [`TicketMachineI`](https://github.com/matsim-org/matsim-libs/blob/01ef87abd7e3faca035ead42c3ba27c75d98adca/contribs/minibus/src/main/java/org/matsim/contrib/minibus/fare/TicketMachineI.java).

Do you have any advice, please? Or can you point me to code where I can navigate through?

**Comment by JWJoubert** [comment]
Hm, any help @neuma or @vsp-gleich? I just need a pointer on how to (programmatically) get a "money event" to a driver/operator's `Stagecontainer` maybe?

**Comment by vsp-gleich** [comment]
Hi @JWJoubert , 

according to the git history the code example does not work since January 2017. I cannot give any better advice than what it is saying, copy and adapt your own `PModule`. That does not seem very hard at first glance, since this class only plugs other classes together and you only have to add or replace one with your own implementation. That does not mean you necessarily have to use a different `PModule`, there might be other options (start on a branch / configure via PConfig) if this is to end up in matsim-libs.

For the `PersonMoneyEvent`: For now those events only hit normal agents, i.e. the passengers. Recently I added a field transactionPartner so one might think of processing that event on the transactionPartner's side, too, but that would be a major new feature with conceptual debates (What kind of object is the transactionPartner? For now it's basically only a free text field).
Those `PersonMoneyEvents` can be handled like all other events with a `PersonMoneyEventHandler`. So you could try to add a `PersonMoneyEventHandler` that catches all money events for your transit driver agent and associate it with your minibus operator, e.g. by listening to `TransitDriverStartsEvents` which gives you the transit line id and that transit line id is the minibus operator's name (check for other transit drivers operating for normal pt from schedule). `StageContainer` seems the wrong place to add this to me since it currently tracks one single passenger whereas the toll information is related to the vehicle. `OperatorCostCollectorHandler` collects costs by vehicle. Therefore it seems a better choice. Maybe add a `PersonMoneyEventHandler` to it on a branch.

I think one important point is that as far as I remember minibus always used shortest path with freespeed travel time routing with the assumption that minibuses drive out car traffic. That also means minibus cannot react to tolls, at most some lines are economically less successful and there might be indirect adaption via the removal or addition of stopsToBeServed which then imply a different freespeed travel time route. But this is an issue you might have to address to see minibus react.

**Comment by kainagel** [comment]
(1) For _scoring_, events are only passed on to persons that are mentioned in the events.  Otherwise, it becomes computationally to expensive.

(2) All other events handlers get all events.

If in your case the bus driver is a person, and you want that that person receives the money event, the easiest I can think of is to throw a second money event, with the same amount but the other sign, and send that to the driver.

Alternatively, what Gregor says would make sense, which is that we say that money events are potentially two-sided things, and so need to be sent to two persons.  However, I think that he is right, and setting this up would take time.  Far easier to just through the money event twice.

(It would be ok with me if code in matsim-libs is changed towards sending the money event twice ... I think that it is a clean solution to the problem we have here, and will not do harm to pre-existing other code.)

**Comment by JWJoubert** [comment]
Thanks, @vsp-gleich & @kainagel. I am currently getting some behavioural responses but the mechanism seems counter-intuitive. If fines go up (either the probability of being caught goes up or the size of the actual fine increases), then we observe fewer and fewer routes and taxi vehicles.

But, in practice, these fines are "spot fines", so they are paid by the driver (from his fare revenue), not by the vehicle owner. And I am just curious as to the impact if we implement it more aligned with what happens in practice if it might affect the outcome/behaviour.

> If in your case the bus driver is a person, and you want that person to receive the money event, the easiest I can think of is to throw a second money event, with the same amount but the other sign, and send that to the driver.

Thanks, I will look into this. I am not sure if the (bus) driver's cost (`PersonMoneyEvents`) actually affects the profitability of a driver's route. I'll run a few small cases and try and figure it out.

**Comment by vsp-gleich** [comment]
> But, in practice, these fines are "spot fines", so they are paid by the driver (from his fare revenue), not by the vehicle owner. And I am just curious as to the impact if we implement it more aligned with what happens in practice if it might affect the outcome/behaviour.

I'm not sure I understand the difference that makes in practice. As far as I understand it is the minibus operator in matsim (representing the vehicle owner in real life) who scores and replans his routes. In matsim the driver of a minibus is nothing else than a normal transit driver (like scheduled pt) with no decision logic whatsoever, he simply drives the minibus along a predefined fixed route. So that transit driver will not react to any `PersonMoneyEvent`.

> Thanks, I will look into this. I am not sure if the (bus) driver's cost (`PersonMoneyEvents`) actually affects the profitability of a driver's route. I'll run a few small cases and try and figure it out.

As far as I understand the transit driver cannot react (see above).

**Comment by JWJoubert** [comment]
Noted, thanks @vsp-gleich. 
>  it is the minibus operator in matsim (representing the vehicle owner in real life) who scores and replans his routes.

That I understand. Let me rephrase: the "hack" I do right now is to affect the operators available budget directly, instead of aggregating it via the `OperatorCostCollectorHandler`. Maybe it has the same effect in the end. I will have to investigate. Even if it has the same net effect, I believe the _better approach_ would be the one that is more intuitive and closer to reality.

**Comment by vsp-gleich** [comment]
`OperatorCostCollectorHandler` calculates costs per vehicle, the lowest level of aggregation, and those costs are aggregated to `TransitRoutes` later. As far as I understand then those routes are the level of aggregation at which the operator decides (add or remove vehicles, copy and adapt the route or remove the route). If you deduct tolls directly from the operator elsewhere you have to ensure those costs are associated with the route otherwise the operator might remove vehicles from all routes instead of only the one with the high tolls.

**Comment by JWJoubert** [comment]
> otherwise the operator might remove vehicles from all routes instead of only the one with the high tolls.

Exactly! And these are the unintended consequences I'm trying to avoid. I'll work on a branch and maybe ask you to review the changes through a pull request.

**Comment by vsp-gleich** [comment]
Sure, I can have a look.
